doc/man7/ossl-guide-libraries-introduction.pod

```diff
@@ -0,0 +1,319 @@
+using the error stack you can find out information such as a reason code/string
+for the error as well as the exact file and source line within OpenSSL that
+emitted the error.
+
+OpenSSL supplies a set of error handling functions to query the error stack. See
+L<ERR_get_error(3)> for information about the functions available for querying
+error data. Also see L<ERR_print_errors(3)> for information on some simple
+helper functions for printing error data. Finally look at L<ERR_clear_error(3)>
+for how to clear old errors from the error stack.
+
+=head1 OPENSSL PROVIDERS
+
+OpenSSL comes with a set of providers.
+
+The algorithms available in each of these providers may vary due to build time
+configuration options. The L<openssl-list(1)> command can be used to list the
+currently available algorithms.
+
+The names of the algorithms shown from L<openssl-list(1)> can be used as an
+algorithm identifier to the appropriate fetching function. Also see the provider
+specific manual pages linked below for further details about using the
+algorithms available in each of the providers.
+
+As well as the OpenSSL providers third parties can also implement providers.
+For information on writing a provider see L<provider(7)>.
+
+=head2 Default provider
+
+The default provider is built-in as part of the F<libcrypto> library and
+contains all of the most commonly used algorithm implementations. Should it be
+needed (if other providers are loaded and offer implementations of the same
+algorithms), the property query string "provider=default" can be used as a
+search criterion for these implementations.  The default provider includes all
+of the functionality in the base provider below.
+
+If you don't load any providers at all then the "default" provider will be
+automatically loaded. If you explicitly load any provider then the "default"
+provider would also need to be explicitly loaded if it is required.
+
+See L<OSSL_PROVIDER-default(7)>.
+
+=head2 Base provider
+
+The base provider is built in as part of the F<libcrypto> library and contains
+algorithm implementations for encoding and decoding for OpenSSL keys.
+Should it be needed (if other providers are loaded and offer
+implementations of the same algorithms), the property query string
+"provider=base" can be used as a search criterion for these implementations.
+Some encoding and decoding algorithm implementations are not FIPS algorithm
+implementations in themselves but support algorithms from the FIPS provider and
+are allowed for use in "FIPS mode". The property query string "fips=yes" can be
+used to select such algorithms.
+
+See L<OSSL_PROVIDER-base(7)>.
+
+=head2 FIPS provider
+
+The FIPS provider is a dynamically loadable module, and must therefore
+be loaded explicitly, either in code or through OpenSSL configuration
+(see L<config(5)>). It contains algorithm implementations that have been
+validated according to the FIPS 140-2 standard. Should it be needed (if other
```

Comment:
Is it Ok to mention 140-2 next to FIPS? It will require us not to forget to make an update here every time we get next certification, e.g. 140-3.

Reply:
FIPS 140-2 or FIPS 1403-3 standards. This should be ok for many many years to come..

Reply:
I just made this say "validated according to FIPS standards" and avoided specific mention of 140-2/140-3.